include/deal.II/grid/tria_faces.h

```diff
@@ -40,100 +40,57 @@ namespace internal
-       * Constructor. This constructor is deleted to prevent the use of this
-       * template, as only the specializations should be used
+       * Constructor.
        */
-      TriaFaces() = delete;
-    };
-
+      TriaFaces(unsigned int dim);
 
-
-    /**
-     * Faces only have a meaning in <tt>dim@>=1</tt>. In <tt>dim=1</tt> they
-     * are vertices, which are handled differently, so only for
-     * <tt>dim@>=2</tt> the use of TriaFaces is reasonable, for <tt>dim=1</tt>
-     * the class is empty.
-     */
-    template <>
-    class TriaFaces<1>
-    {
-    public:
       /**
-       * Determine an estimate for the memory consumption (in bytes) of this
-       * object. Of course this returns 0.
+       * Default constructor for Boost::serialization.
        */
-      std::size_t
-      memory_consumption() const;
+      TriaFaces() = default;
 
       /**
-       * Read or write the data of this object to or from a stream for the
-       * purpose of serialization
+       * Dimension of the underlying triangulation.
        */
-      template <class Archive>
-      void
-      serialize(Archive &ar, const unsigned int version);
-    };
-
-    /**
-     * In <tt>dim=2</tt> the cells are quads, the faces accordingly are lines.
-     */
-    template <>
-    class TriaFaces<2>
-    {
-    public:
-      /**
-       * The TriaObject containing the data of lines.
-       */
-      TriaObjects<TriaObject<1>> lines;
+      unsigned int dim;
 
-    public:
       /**
-       * Determine an estimate for the memory consumption (in bytes) of this
-       * object.
+       * The TriaObject containing the data of quads.
+       *
+       * @note Used only for dim=3.
        */
-      std::size_t
-      memory_consumption() const;
+      TriaObjects quads;
 
       /**
-       * Read or write the data of this object to or from a stream for the
-       * purpose of serialization
+       * Orientation of each line of each quad.
+       *
+       * @note Used only for dim=3.
        */
-      template <class Archive>
-      void
-      serialize(Archive &ar, const unsigned int version);
-    };
+      std::vector<char> quads_line_orientations;
 
-    /**
-     * In <tt>dim=3</tt> the cells are hexes, the faces accordingly are quads.
-     * In addition to that we also have to enable the storage of lines.
-     */
-    template <>
-    class TriaFaces<3>
-    {
-    public:
       /**
-       * The TriaObject containing the data of quads.
+       * The TriaObject containing the data of lines.
+       *
+       * @note Used only for dim>1.
        */
-
-      TriaObjectsQuad3D quads;
+      TriaObjects lines;
 
       /**
-       * The TriaObject containing the data of lines.
+       * Reserve space for line_orientations.
+       *
+       * @note Used only for dim=3.
        */
-      TriaObjects<TriaObject<1>> lines;
+      void
+      reserve_space(const unsigned int new_quads_in_pairs,
+                    const unsigned int new_quads_single = 0);
```

Comment:
Can you document what the two arguments do?

Reply:
I would if I knew what they  mean in `internal::TriangulationImplementation::TriaObjects::reserve_space()` ;)

Reply:
My suggestion is to postpone this issue. I will dig into `Triangulation` and the internal data structures more deeply in the next weeks. The internal functions might change and/or I might understand the comments then!

Reply:
In the context of adding children, we used to store all `2^d` children of one cell consecutively and each cell only stored a pointer to the first of its children. But then along came anisotropic refinement where we may have only 2 children, so we now store them in blocks of 2 and store `2^(d-1)` pointers to blocks of 2 children.

I would not be surprised if this is a similar thing: Where we typically store two new faces consecutively, but then there are occasions where we need a single one. That's sort of my educated guess from the context and the names, but I really have no idea whether that's true.